### PR TITLE
[FLOC-2806] A test for containers of private registry images

### PR DESCRIPTION
**Pull request by wallrj**
Fixes: https://clusterhq.atlassian.net/browse/FLOC-2806

http://build.clusterhq.com/boxes-flocker?branch=test-private-registry-FLOC-2806&num_builds=2

* First build is against the old docker-py-0.7.1 and should fail to pull the image. (had to specify `insecure_registry=True,` when pushing to work around bug in docker-py.)
* Second build is against new docker-py-1.3.1 and should work.

### CODE REVIEW

File: flocker/node/functional/test_docker.py [hunk 1]
```diff
@@ -170,6 +170,59 @@ def started(_):
...
+        address_user = socket.socket()
+        self.addCleanup(address_user.close)
+        address_user.bind(('', 0))
+        used_address = address_user.getsockname()
+        address_user.close()
+        registry_name = random_name(self)
+        registry_port = used_address[1]
+        registry_starting = self.start_container(
+            unit_name=registry_name,
+            image_name='registry:2',
```
Comment (wallrj) [inline flocker/node/functional/test_docker.py]: I should probably add the registry image to the  list of images pre-pulled by the build slave.

Comment (wallrj) [inline flocker/node/functional/test_docker.py]: Also add a link to https://docs.docker.com/registry/deploying/

File: flocker/node/functional/test_docker.py [hunk 2]
```diff
@@ -170,6 +170,59 @@ def started(_):
...
+    def test_private_registry_image(self):
+        """
+        ``DockerClient.add`` can start containers based on an image from a
+        private registry.
+        """
+        address_user = socket.socket()
+        self.addCleanup(address_user.close)
+        address_user.bind(('', 0))
+        used_address = address_user.getsockname()
+        address_user.close()
```
Comment (wallrj) [inline flocker/node/functional/test_docker.py]: Look for an existing helper function...there must be one, right?

File: flocker/node/functional/test_docker.py [hunk 3]
```diff
@@ -170,6 +170,59 @@ def started(_):
...
+            unit_name=registry_name,
+            image_name='registry:2',
+            ports=[
+                PortMap(
+                    internal_port=5000,
+                    external_port=registry_port
+                ),
+            ]
+        )
+        repository = '127.0.0.1:{}/clusterhq/testimage'.format(
```
Comment (wallrj) [inline flocker/node/functional/test_docker.py]: Maybe use a `random_name` here.

File: flocker/node/functional/test_docker.py [hunk 4]
```diff
@@ -170,6 +170,59 @@ def started(_):
...
+            registry_port
+        )
+        tag = '123'
+
+        def push_image(client):
+            client._client.tag(
```
Comment (wallrj) [inline flocker/node/functional/test_docker.py]: Maybe instantiate a separate `TimeoutClient` instead of using the private client here.

File: flocker/node/functional/test_docker.py [hunk 5]
```diff
@@ -170,6 +170,59 @@ def started(_):
...
+                image='busybox',
```
Comment (wallrj) [inline flocker/node/functional/test_docker.py]: Could use `dockerimage_builder` rather than re-tagging an existing image. Not sure if it makes much difference.

File: flocker/node/functional/test_docker.py (hunk 4 above)
Comment (wallrj) [inline flocker/node/functional/test_docker.py]: Yeah, I definitely need to do that:

```
[ERROR]
Traceback (most recent call last):
  File "/srv/buildslave/flocker-centos-7/venv/lib/python2.7/site-packages/flocker/node/functional/test_docker.py", line 201, in push_image
    client._client.tag(
exceptions.AttributeError: 'DockerClient' object has no attribute 'tag'

flocker.node.functional.test_docker.NamespacedDockerClientTests.test_private_registry_image
-------------------------------------------------------------------------------
Ran 2087 tests in 812.477s

FAILED (skips=102, expectedFailures=8, errors=1, successes=1976)
```

**Comment by adamtheturtle** [comment]
A failing test which is not mentioned in the comments is:

```
[ERROR]
Traceback (most recent call last):
  File "/srv/buildslave/flocker-coverage/venv/local/lib/python2.7/site-packages/flocker/node/functional/test_docker.py", line 208, in push_image
    tag=tag,
  File "/srv/buildslave/flocker-coverage/venv/local/lib/python2.7/site-packages/docker/client.py", line 684, in push
    self._raise_for_status(response)
  File "/srv/buildslave/flocker-coverage/venv/local/lib/python2.7/site-packages/docker/clientbase.py", line 104, in _raise_for_status
    raise errors.APIError(e, response, explanation=explanation)
docker.errors.APIError: 500 Server Error: Internal Server Error ("Invalid registry endpoint "http://127.0.0.1:42499/v1/". HTTPS attempt: Get https://127.0.0.1:42499/v1/_ping: read tcp 127.0.0.1:42499: connection reset by peer. HTTP attempt: Get http://127.0.0.1:42499/v1/_ping: read tcp 127.0.0.1:42499: connection reset by peer")

flocker.node.functional.test_docker.GenericDockerClientTests.test_private_registry_image
```

http://build.clusterhq.com/builders/flocker-coverage/builds/8228/steps/trial/logs/problems

File: flocker/node/functional/test_docker.py (hunk 3 above)
Comment (adamtheturtle) [inline flocker/node/functional/test_docker.py]: Why?

File: flocker/node/functional/test_docker.py (hunk 2 above)
Comment (adamtheturtle) [inline flocker/node/functional/test_docker.py]: Why is this here when it is also in the test cleanup?

File: flocker/node/functional/test_docker.py (hunk 3 above)
Comment (adamtheturtle) [inline flocker/node/functional/test_docker.py]: I think that this is an `image_location` (or at least, it isn't just a repository)

File: flocker/node/functional/test_docker.py [hunk 6]
```diff
@@ -170,6 +170,59 @@ def started(_):
...
+            )
+            client._client.push(
+                repository=repository,
+                tag=tag,
+            )
+            client._client.remove_image(
+                image=repository + ':' + tag,
+            )
+            return client
+        pushing_image = registry_starting.addBoth(push_image)
```
Comment (adamtheturtle) [inline flocker/node/functional/test_docker.py]: Would we not want the test to fail if starting a container fails?

**Comment by adamtheturtle** [comment]
Thanks @wallrj , I think that there are enough comments (mostly yours) to get on with. Please resubmit when the tests are passing and the comments have been addressed.

File: flocker/node/functional/test_docker.py (hunk 2 above)
Comment (wallrj) [inline flocker/node/functional/test_docker.py]: It was copy pasted code. I've found a helper function for this.

File: flocker/node/functional/test_docker.py (hunk 1 above)
Comment (wallrj) [inline flocker/node/functional/test_docker.py]: Done and Done: https://clusterhq.atlassian.net/browse/FLOC-2828

File: flocker/node/functional/test_docker.py (hunk 3 above)
Comment (wallrj) [inline flocker/node/functional/test_docker.py]: > Why?
Because if the removal of the local image fails for some reason, it'll be obvious which test left the image behind and so that subsequent test runs don't use the local image instead of pulling from the temporary repository.

> I think that this is an image_location (or at least, it isn't just a repository)
I agree it's confusing, but repository is the name of the attribute we've used in `flocker.control.DockerImage` and I've re-used that record now.

That record should probably grow separate repository, name and tag attributes which will be part of FLOC-246 now linked from the new test.

File: flocker/node/functional/test_docker.py (hunk 5 above)
Comment (wallrj) [inline flocker/node/functional/test_docker.py]: Done. Because having multiple tags of busybox seems to cause problems when removing the tags.

File: flocker/node/functional/test_docker.py (hunk 6 above)
Comment (wallrj) [inline flocker/node/functional/test_docker.py]: Yes, should be addCallback

**Comment by wallrj** [comment]
Putting this up for review again.

I struggled to get the tests to pass reliably on my local vagrant dev box. But that might be because Docker is broken somehow. 

I've:
 * Addressed all mine and Adam's comments
 * Switched to using random names for images to ensure that the image is being pulled from the private registry rather than using a locally cached image.

Some problems / confusion I encountered: 

 * Decided *not* to use DockerImageBuilder once I realised why the locally created tags couldn't be deleted because docker refuses to remove tags that reference the same image in different repositories.
 * Tried then abandoned the idea of limiting the length of the image repository/name combination to 30 chars as https://github.com/docker/docker/issues/10392 suggests that that is the limit. There doesn't seem to be such a limit.

Hopefully the tests will pass reliably now.

File: flocker/node/functional/test_docker.py [hunk 7]
```diff
@@ -170,6 +172,89 @@ def started(_):
...
+                image='openshift/busybox-http-app',
+                repository=private_image.repository,
+                tag=private_image.tag,
+            )
+            # Push to the local registry.
+            client.push(
+                repository=private_image.repository,
+                tag=private_image.tag,
+            )
+            # Remove the local tag of the random image
```
Comment (itamarst) [inline flocker/node/functional/test_docker.py]: This step seems wrong. Why are you deleting the thing you just pushed?

Comment (itamarst) [inline flocker/node/functional/test_docker.py]: And indeed the tests are failing. Delete this and see if it fixes it?

**Comment by itamarst** [comment]
Test looks good, presuming it actually passes. So if all you had to do is delete that unnecessary-seeming remove_image then merge once tests are green. If you need to do something more significant resubmit if you feel it needs another pass.

File: flocker/node/functional/test_docker.py (hunk 7 above)
Comment (wallrj) [inline flocker/node/functional/test_docker.py]: I delete the local image / tag so that it gets pulled from the private repository before the container starts in the next step.

**Comment by wallrj** [comment]
I wonder if the Ubuntu builder fails because the old version of Docker does have a limit on image names.

File: flocker/node/functional/test_docker.py (hunk 7 above)
Comment (itamarst) [inline flocker/node/functional/test_docker.py]: Oh hm. OK I can see that.

**Comment by wallrj** [comment]
Nope, that didn't work. I guess I need to fire up an Ubuntu 14.04  test machine and dig into the docker logs.

**Comment by wallrj** [comment]
Aha. READ THE DOCS Richard. 
 * https://docs.docker.com/registry/deploying/

> You obviously need to install Docker (remember you need Docker version 1.6.0 or newer).

We're installing Docker 1.3.3 on Ubuntu

```
(2806)root@acceptance-test-richardw-0:~/flocker# docker version
Client version: 1.3.3
Client API version: 1.15
Go version (client): go1.2.1
Git commit (client): d344625
OS/Arch (client): linux/amd64
Server version: 1.3.3
Server API version: 1.15
Go version (server): go1.2.1
Git commit (server): d344625
```

I'll try upgrading to latest docker and re-running the tests.

**Comment by wallrj** [comment]
Yeah the tests pass when I upgrade Docker

```
(2806)root@acceptance-test-richardw-0:~/flocker# docker info
Containers: 3
Images: 292
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 298
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-58-generic
Operating System: Ubuntu 14.04.2 LTS
CPUs: 8
Total Memory: 7.793 GiB
Name: acceptance-test-richardw-0
ID: CJYO:QDY7:72DB:WVAD:HRHS:ZDTR:5G6P:5UPB:GYKV:MGUA:ASDO:CBHH
(2806)root@acceptance-test-richardw-0:~/flocker# docker version
Client version: 1.7.1
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 786b29d
OS/Arch (client): linux/amd64
Server version: 1.7.1
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 786b29d
OS/Arch (server): linux/amd64
(2806)root@acceptance-test-richardw-0:~/flocker# git diff
diff --git a/flocker/node/functional/test_docker.py b/flocker/node/functional/test_docker.py
index 159fbe7..978ceaa 100644
--- a/flocker/node/functional/test_docker.py
+++ b/flocker/node/functional/test_docker.py
@@ -217,6 +217,7 @@ class GenericDockerClientTests(TestCase):
         )
 
         def tag_and_push_image(ignored):
+            import pdb; pdb.set_trace()
             client = Client()
             # The image will normally have been pre-pulled on build slaves, but
             # may not already be available when running tests locally.
@@ -248,6 +249,7 @@ class GenericDockerClientTests(TestCase):
         pushing_image = registry_listening.addCallback(tag_and_push_image)
 
         def start_private_image(ignored):
+            import pdb; pdb.set_trace()
             return self.start_container(
                 unit_name=random_name(self),
                 image_name=private_image.full_name,
(2806)root@acceptance-test-richardw-0:~/flocker# git reset --hard
HEAD is now at 1289167 Limit the image name length to 30 in an attempt to get Ubuntu builders to pass
(2806)root@acceptance-test-richardw-0:~/flocker# trial --rterrors flocker.node.functional.test_docker.NamespacedDockerClientTests.test_private_registry_image
flocker.node.functional.test_docker
  NamespacedDockerClientTests
    test_private_registry_image ...                                        [OK]

-------------------------------------------------------------------------------
Ran 1 tests in 9.468s

PASSED (successes=1)
(2806)root@acceptance-test-richardw-0:~/flocker# trial --rterrors flocker.node.functional.test_docker.NamespacedDockerClientTests.^Cst_private_registry_image
(2806)root@acceptance-test-richardw-0:~/flocker# trial --rterrors flocker.node.functional.test_docker.{Generic,Namespaced}DockerClientTests.test_private_registry_image
flocker.node.functional.test_docker
  GenericDockerClientTests
    test_private_registry_image ...                                        [OK]
  NamespacedDockerClientTests
    test_private_registry_image ...                                        [OK]

-------------------------------------------------------------------------------
Ran 2 tests in 16.948s

PASSED (successes=2)
```

I can skip this test on Ubuntu and / or mark the issue as blocked by the Docker upgrade.
 * https://clusterhq.atlassian.net/browse/FLOC-2524

**Comment by wallrj** [comment]
Actually skip when the docker version is `<1.6.0`.

**Comment by wallrj** [comment]
Centos-7 builder runs the test and it passes:
 * http://build.clusterhq.com/builders/flocker-centos-7/builds/5993/steps/trial/logs/stdio

Ubuntu-14.04 builder now skips the test:
 * http://build.clusterhq.com/builders/flocker-ubuntu-14.04/builds/5664/steps/trial/logs/stdio

```
[SKIPPED]
Minimum required Docker version: 1.6.0. Actual Docker version: 1.3.3. Details: This test uses the registry:2 image which requires Docker-1.6.0 or newer. See https://docs.docker.com/registry/deploying/ for details.

flocker.node.functional.test_docker.GenericDockerClientTests.test_private_registry_image
flocker.node.functional.test_docker.NamespacedDockerClientTests.test_private_registry_image
===============================================================================
```

File: flocker/node/functional/test_docker.py [hunk 8]
```diff
@@ -170,6 +174,97 @@ def started(_):
...
+
+        Docker can pull from a private registry without any TLS configuration
+        as long as it's running on the local host.
+        """
+        registry_name = random_name(self)
+        registry_port = find_free_port()[1]
+        repository = '127.0.0.1:{}'.format(registry_port)
+        # Total image name seems to be limited to 30 bytes for some versions of
+        # Docker: https://github.com/docker/docker/issues/10392
+        name = random_name(self).lower()[-(29 - len(repository)):]
```
Comment (jml) [inline flocker/node/functional/test_docker.py]: I _think_ this would be simpler and more correct if we did the truncation when we construct the `DockerImage` below, because it could be expressed as `repository=(repository + '/' + name)[:30]`.

File: flocker/node/testtools.py [hunk 9]
```diff
@@ -52,6 +55,42 @@ def docker_accessible():
     ))
 
 
+def require_docker_version(minimum_docker_version, message='N/A'):
```
Comment (jml) [inline flocker/node/testtools.py]: :question: Might as well make `message` compulsory for now, since it's provided in the only use.

File: flocker/node/functional/test_docker.py [hunk 10]
```diff
@@ -170,6 +174,97 @@ def started(_):
         d.addCallback(started)
         return d
 
+    @require_docker_version(
+        '1.6.0',
+        'This test uses the registry:2 image '
+        'which requires Docker-1.6.0 or newer. '
+        'See https://docs.docker.com/registry/deploying/ for details.'
+    )
```
Comment (jml) [inline flocker/node/functional/test_docker.py]: This is nice, thanks.

**Comment by jml** [comment]
Minor change requested. If you make it and there are no changes required to make the tests pass, then feel free to merge.

Otherwise, please resubmit for review.

File: flocker/node/testtools.py (hunk 9 above)
Comment (wallrj) [inline flocker/node/testtools.py]: Done.

File: flocker/node/functional/test_docker.py (hunk 8 above)
Comment (wallrj) [inline flocker/node/functional/test_docker.py]: The image name length restriction was a red herring (I think).
I removed it.
